File Formats/Images/UCG Image Format version 3.md

```diff
@@ -0,0 +1,291 @@
+# **Universal Compressed Graphics file format version 3**
+
+## Contents
+1. **Abstract**
+2. **Quick Overview**
+3. **Introduction & Purpose**
+4. **Specification** <br>
+	4.1 Definitions <br>
+	&emsp;4.1.1 Byte <br>
+	&emsp;4.1.2 Word <br>
+	&emsp;4.1.3 ColorValue <br>
+	&emsp;4.1.4 LengthValue <br>
+	&emsp;4.1.5 HuffmanTree <br>
+	&emsp;4.1.6 PixelData <br>
+5. **Explanation** <br>
+	5.1 Signature <br>
+	5.2 version <br>
+	5.3 Flags <br>
+	5.4 Width <br>
+	5.5 Height <br>
+	5.6 ColorHuffmanTree <br>
+	5.7 LengthHuffmanTree <br>
+	5.8 PixelData <br>
+	5.9 Huffman Codes <br>
+6. **See Also**
+
+## 1. Abstract
+This document defines and explains a file format for storing images and graphics, specificated for the usage with ComputerCraft and CraftOS 2.
+
+## 2. Quick Overview
+
+| Name       | Value             |
+| ---------- | ----------------- |
+| Type       | Image file format |
+| MIME       | ``image/ucg``     |
+| Extensions | ``.ucg``          |
```

Comment:
Semantic versioning is not suitable for file formats. First, file formats are tested very thoroughly, so small patches are unwanted and don't really appear. Updates in general are unwanted for file formats. PNG has not been updated since 1999, although there are many better compression algorithms out there now than the one PNG uses; and the reasons is backwards compatibility.

Also, when you use a two value version tag like ``MAJOR.MINOR``, and you want to compare if yours is greater than another version, you would just do
```lua
if 256*major1 + minor1 > 256*major2 + minor2 then
  -- ...
end
```
To get the value of a word, you do ``256*firstbyte + secondbyte``. Notice the similarity? In practice, you would just read the version of a file as a word instead of a byte. I decided, I don't need this one byte more just to know if the decoder is able to read the file or not.


This is not the very first version of this file format. There were versions before that oeed and me used, and there were versions of libucg before that used these UCG versions. If I release libucg on the forums, I also release the older versions. Of course, noone's gonna use these, but I decided I don't fall back to version 1, because it'd miss the point of versioning.